### ISSUE TITLE
Using exception args in Keyword shows a type error

### ISSUE THREAD

**Issue by christopheradams** [jeremyjh/dialyxir#51]
Using the `defexception` macro gives us an overridable `exception/1` function whose typespec is:

`@spec exception(Keyword.t) :: Exception.t`

 However, treating the argument as a Keyword list results in a type error (even though the code compiles and works):

``` elixir
defmodule MyAppError do
  defexception [:message]

  def exception(args) when is_list(args) do
    message = Keyword.get(args, :message)
    %MyAppError{message: message}
  end
end
```

```
my_app.ex:2: Type specification 'Elixir.MyAppError':exception('Elixir.Keyword':t()) -> 'Elixir.Exception':t()
    ; ('Elixir.String':t()) -> 'Elixir.Exception':t() is a supertype of the success typing: 
'Elixir.MyAppError':exception([{atom(),_}]) -> #{'__exception__':='true', '__struct__':='Elixir.MyAppError', 'message':=_}
```

If I first pass `args` to `Keyword.new/1` or access it with `List.keyfind/4` the error goes away:

``` elixir
  def exception(args) when is_list(args) do
    message = args |> Keyword.new |> Keyword.get(:message)
    {:message, message} = List.keyfind(args, :message, 0)
    %MyAppError{message: message}
  end

```

Any idea what's going on here?

**Comment by davidsulc** [comment]
I'm running into a similar issue, although (it seems) for a different reason (I'm not accessing `:message`).

My code:

```
defmodule SimpleTranslator.Errors do
  defmodule ConfigurationError do
    defexception [:message]

    def message(exception), do: exception.message

    @spec exception(nonempty_list(atom)) :: Exception.t
    def exception([model: model, association: association]) do
      msg = "translations for '#{model}' were misconfigured: association '#{association}' doesn't exist"
      %__MODULE__{message: msg}
    end
  end

  defmodule TranslationNotFoundError do
    defexception [:message]

    def message(exception), do: exception.message

    @spec exception(nonempty_list(atom)) :: Exception.t
    def exception([locale: locale, association: association]) do
      msg = "association '#{association}' does not contain translation for locale '#{locale}'"
      %__MODULE__{message: msg}
    end

    def exception([locale: locale]) do
      msg = "translation not found for locale '#{locale}'"
      %__MODULE__{message: msg}
    end
  end
end
```

The errors I get in dialyxir:

```
Compiling 2 files (.ex)
Starting Dialyzer
dialyzer --no_check_plt --plt /home/david/.dialyxir_core_19_1.3.2.plt -Wunmatched_returns -Werror_handling -Wrace_conditions -Wunderspecs /home/david/Documents/dev/simple_translator/_build/dev/lib/simple_translator/ebin
  Proceeding with analysis...
errors.ex:3: Type specification 'Elixir.SimpleTranslator.Errors.ConfigurationError':exception([atom(),...]) -> 'Elixir.Exception':t()
    ; ('Elixir.Keyword':t()) -> 'Elixir.Exception':t()
    ; ('Elixir.String':t()) -> 'Elixir.Exception':t() is a supertype of the success typing: 'Elixir.SimpleTranslator.Errors.ConfigurationError':exception([{'association',atom() | bitstring() | fun() | pid() | port() | reference() | maybe_improper_list() | number() | tuple() | map()} | {'model',atom() | bitstring() | fun() | pid() | port() | reference() | maybe_improper_list() | number() | tuple() | map()},...]) -> #{'__exception__':='true', '__struct__':='Elixir.SimpleTranslator.Errors.ConfigurationError', 'message':=<<_:64,_:_*8>>}
errors.ex:17: Type specification 'Elixir.SimpleTranslator.Errors.TranslationNotFoundError':exception([atom(),...]) -> 'Elixir.Exception':t()
    ; ('Elixir.Keyword':t()) -> 'Elixir.Exception':t()
    ; ('Elixir.String':t()) -> 'Elixir.Exception':t() is a supertype of the success typing: 'Elixir.SimpleTranslator.Errors.TranslationNotFoundError':exception([{'association',atom() | bitstring() | fun() | pid() | port() | reference() | maybe_improper_list() | number() | tuple() | map()} | {'locale',atom() | bitstring() | fun() | pid() | port() | reference() | maybe_improper_list() | number() | tuple() | map()},...]) -> #{'__exception__':='true', '__struct__':='Elixir.SimpleTranslator.Errors.TranslationNotFoundError', 'message':=<<_:64,_:_*8>>}
loader.ex:25: Invalid type specification for function 'Elixir.SimpleTranslator.Loader':'fetch_translation!'/3. The success typing is (#{'__struct__':=atom(), 'id':=_, atom()=>_},atom() | bitstring() | fun() | pid() | port() | reference() | maybe_improper_list() | number() | tuple() | map(),atom()) -> any()
 done in 0m1.38s
done (warnings were emitted)
```

**Comment by christopheradams** [comment]
> I'm running into a similar issue, although (it seems) for a different reason (I'm not accessing :message).

@davidsulc it looks like the same underlying issue. the only difference is that you're pattern matching on the keyword list in the function head, rather than using `Keyword.get`.

**Comment by jeremyjh** [comment]
Hi, I could not reproduce this with Elixir 1.3.4 or 1.40. I think perhaps it was fixed in [44ef53e](https://github.com/elixir-lang/elixir/commit/44ef53ec2b1f49d86a35a0b33d6c1209fadd3cc8). Is this an issue on 1.3.3 or later?

**Comment by christopheradams** [comment]
Looks good now. Thanks!
